contracts/AnchoredPriceView.sol

```diff
@@ -0,0 +1,331 @@
+        cUsdcAddress = tokens_.cUsdcAddress;
+        cDaiAddress = tokens_.cDaiAddress;
+        cRepAddress = tokens_.cRepAddress;
+        cWbtcAddress = tokens_.cWbtcAddress;
+        cBatAddress = tokens_.cBatAddress;
+        cZrxAddress = tokens_.cZrxAddress;
+        cSaiAddress = tokens_.cSaiAddress;
+        cUsdtAddress = tokens_.cUsdtAddress;
+    }
+
+    /**
+     * @notice Primary entry point to post and recalculate prices
+     * @dev We let anyone pay to post anything, but only prices from configured source will be stored in the view
+     * @param messages The messages to post to the oracle
+     * @param signatures The signatures for the corresponding messages
+     */
+    function postPrices(bytes[] calldata messages, bytes[] calldata signatures, string[] calldata symbols) external {
+        require(messages.length == signatures.length, "messages and signatures must be 1:1");
+
+        // Save the prices
+        for (uint i = 0; i < messages.length; i++) {
+            priceData.put(messages[i], signatures[i]);
+        }
+
+        // load usdc for using in loop to convert anchor prices to dollars
+        uint256 usdcPrice = anchor.getUnderlyingPrice(cUsdcAddress);
+
+        // Try to update the view storage
+        for (uint i = 0; i < symbols.length; i++) {
+            string memory symbol = symbols[i];
+            address tokenAddress = getCTokenAddress(symbol);
+            uint256 sourcePrice = priceData.getPrice(source, symbol);
+            uint256 anchorPrice = getAnchorPrice(tokenAddress, usdcPrice);
+
+            if (anchorPrice == 0 || tokenAddress == cUsdcAddress || tokenAddress == cUsdtAddress) {
+                emit PriceGuarded(symbol, sourcePrice, anchorPrice);
+            } else {
+                uint256 anchorRatioMantissa = sourcePrice * 100e16 / anchorPrice;
+                // Only update the view's price if the source price is within a bound, and it is a new median
+                if (anchorRatioMantissa <= upperBoundAnchorRatio && anchorRatioMantissa >= lowerBoundAnchorRatio) {
+                    // only update and emit event if the source price is new, otherwise do nothing
+                    if (_prices[symbol] != sourcePrice) {
+                        _prices[symbol] = sourcePrice;
+                        emit PriceUpdated(symbol, sourcePrice);
+                    }
+                } else {
+                    emit PriceGuarded(symbol, sourcePrice, anchorPrice);
+                }
+            }
+        }
+    }
+
+    /**
+     * @notice Returns price denominated in USD, with 6 decimals
+     * @dev If price was posted by source, return it. Otherwise, return anchor price converted through source ETH price.
+     */
+    function prices(string calldata symbol) external view returns (uint256) {
+        uint256 price = _prices[symbol];
+
+        if (price != 0) {
+            return price;
+        } else {
+            uint256 usdPerEth = _prices["ETH"];
```

Comment:
what is our source failed to post ETH price, should we have a check that price of ETH is not 0?

Reply:
if the source hasn't posted eth at least once, we should not use this view at all and returning 0 for every price is a reasonable result. 

A require statement and reverting may be better though?

Reply:
Yeah, this sounds look a good place to `require`